Make debug_pkl smaller by only emitting unique traces.

qihqi opened the pull request:
Summary:
debug_pkl file inside of pytorch's .pt file consists of a list of SourceRanges. Each SourceRange points to a Source which is a stack track, filename, and start, end numbers. Those are emitted in debug_pkl file as strings.

Since many SourceRange contains similar looking strings, they can be deduped.

The newer format saves a set of unique traces in a tuple, then each SourceRange will save the offset of it's trace w.r.t. position in that tuple. (i.e. manually applying dictionary compression).

Test Plan: unit test

Differential Revision: D33994011

gmagogsfm commented on torch/csrc/jit/serialization/source_range_serialization.cpp:
```diff
@@ -65,10 +91,36 @@ c10::IValue SourceRangeSerializer::serialize_source(
+      text_pos = static_cast<int64_t>(texts_.size());
+      IValue text = s->text();
+      text_to_pos_[s->text()] = text_pos;
+      texts_.push_back(text);
```
Might be cheaper to emplace_back `s_text()` directly

qihqi replied:
done.

gmagogsfm commented on torch/csrc/jit/serialization/source_range_serialization.cpp:
```diff
@@ -65,10 +91,36 @@ c10::IValue SourceRangeSerializer::serialize_source(
-    serialized = c10::ivalue::Tuple::create({"", "", 0});
+    serialized = c10::ivalue::Tuple::create({-1, -1, 0});
   } else {
+    auto text_iter = text_to_pos_.find(s->text());
```
Nit:
rename `text_to_pos_` to `text_to_idx_` because `pos` has another meaning in the context of source range.

gmagogsfm replied:
Another nit:
This chunk of logic might be a good fit for a utility func, so that it is more readable and reusable by the fname caching logic below

qihqi replied:
done.

gmagogsfm commented on torch/csrc/jit/serialization/source_range_serialization.cpp:
```diff
@@ -10,18 +11,26 @@ namespace jit {
   // the serialized form of Source
   c10::IValue serialize(const SourceRange& sr);
 
+  const std::vector<c10::IValue>& textsSaved() {
```
I think naming convention for methods in this file is lower case words connected by underscores.

gmagogsfm commented on torch/csrc/jit/serialization/source_range_serialization.cpp:
```diff
@@ -65,10 +91,36 @@ c10::IValue SourceRangeSerializer::serialize_source(
+
     serialized = c10::ivalue::Tuple::create(
-        {s->text(), s->filename(), (int64_t)s->starting_line_no()});
+        {text_pos, fname_pos, (int64_t)s->starting_line_no()});
```
Looks like this will dedup source code, could we run a one-time manual test on that ads model to see how much saving we can get?

How about the stack trace part in debug_pkl? Could we compress them too? It looks like a lot of the strings have redundancy.

qihqi replied:
> Looks like this will dedup source code, could we run a one-time manual test on that ads model to see how much saving we can get?
I'll get help from the ads folks to see how to do this.

> How about the stack trace part in debug_pkl? Could we compress them too? It looks like a lot of the strings have redundancy.
yes. SourceView object has string field "text", string field "filename". Sometimes the "text" field is stuffed with source code and sometimes stuffed with stack trace. I initially thought it's always source code until I unpicked and printed out what is in there and saw stack trace.

gmagogsfm reviewed (commented):
You mentioned that unit tests are used as test plans, but I don't see a checked in unit test. Is it left out of commit accidentally?

qihqi commented on torch/csrc/jit/mobile/debug_info.cpp:
```diff
@@ -122,25 +122,38 @@ MobileDebugTable::MobileDebugTable(
+      if (ivalues[0].isString() &&
+          kFormatWithStringTable == ivalues[0].toStringRef()) {
+        for (const auto& x : ivalues[1].toTuple()->elements()) {
+          std::cerr << "access, " << x.toStringRef()[0] << std::endl;
```
removed.

qihqi commented on torch/csrc/jit/mobile/debug_info.h:
```diff
@@ -50,6 +51,8 @@ class MobileDebugTable {
       const std::string& top_module_type_name = "ModuleTypeUnknown") const;
   ska::flat_hash_map<int64_t, DebugInfoTuple> callstack_ptr_map_;
+  std::shared_ptr<SourceRangeDeserializer> deserializer_;
+  std::vector<IValue> text_tables_;
```
done.

qihqi commented:
> You mentioned that unit tests are used as test plans, but I don't see a checked in unit test. Is it left out of commit accidentally?

I meant existing unit tests that do save/load sequences. Ex. test_jit

gmagogsfm commented on torch/csrc/jit/serialization/source_range_serialization.cpp:
```diff
@@ -58,17 +86,35 @@ c10::IValue SourceRangeSerializer::serialize(const SourceRange& sr) {
+    serialized = c10::ivalue::Tuple::create({-1, -1, 0});
   } else {
+    int64_t text_pos = store_text_and_get_index(s->text());
+    int64_t fname_pos = -1;
```
I am a bit worried about using -1 as default value because Pickle doesn't have schema or verification anywhere. Could we set 0 as default value and push some string that says "not_available" in there so that a naive consumer of `fname_pos` wouldn't crash by looking up text table by idx without checking?

gmagogsfm commented on torch/csrc/jit/serialization/source_range_serialization.cpp:
```diff
@@ -80,20 +126,27 @@ std::vector<char> SourceRangePickler::pickle(
     const SourceRangeRecords& ranges,
     const SourceRangeTagMap& source_range_tags) {
   std::vector<c10::IValue> ivalues;
+  std::unordered_map<c10::string_view, int> trace_to_pos;
```
Maybe replace `pos` with `idx` too here? Wdyt?

gmagogsfm commented on torch/csrc/jit/frontend/source_range.h:
```diff
@@ -4,37 +4,320 @@
+    return pieces_[index];
+  }
+
+  struct Iterator {
```
Consider renaming the class, otherwise it would become `torch::jit::Iterator` which is too broad for what it covers.

qihqi replied:
it's torch::jit::StringCordView::Iterator sicne it's nested inside the class.

gmagogsfm commented on torch/csrc/jit/frontend/lexer.h:
```diff
@@ -190,7 +190,7 @@ struct TORCH_API SharedParserData {
   // if successful filling in kind, start,and len
   bool match(
-      c10::string_view str,
+      StringCordView str,
```
In theory we shouldn't need to optimize memory foorprint inside `lexer` because its source text should go out of scope once compilation is completed.

Same for other use cases like onnx converter, python tree view etc.

Could we verify if this is the case? If it is, maybe we can hide `StringCordView` as an implementation detail behind `source`?

qihqi replied:
> maybe we can hide StringCordView as an implementation detail behind source?

This can be done if Source implements methods such as substr, find directly; then we forward the call to StringCordView. At that point; Source becomes StringCordView. Then, match need to take Source as input instead of StringCordView (still cannot take c10::string_view as input, and we are really in the same spot).

gmagogsfm commented on torch/csrc/jit/frontend/lexer.h:
```diff
@@ -241,12 +241,12 @@ struct TORCH_API SharedParserData {
     *start = pos;
     // check for a valid number
-    if (isNumber(str, pos, len)) {
+    if (isNumber(str.piece(0), pos, len)) {
```
Maybe `StringCordView` should have a method that's called `only_piece`, which returns `piece(0)` and asserts there is only a single piece?

qihqi replied:
That won't work because if I save a Source then load it again, then its StringCordView is no longer only_piece.
This change is correct if in StringCordView never split a token into 2 pieces.

gmagogsfm commented on torch/csrc/jit/frontend/function_schema_parser.cpp:
```diff
@@ -27,8 +27,8 @@ namespace jit {
-  SchemaParser(const std::string& str)
-      : L(std::make_shared<SourceView>(c10::string_view(str))),
+  explicit SchemaParser(const std::string& str)
+      : L(std::make_shared<Source>(c10::string_view(str))),
```
Does this mean we are copying entire string (within Source) when schema parser is created?

qihqi replied:
No. Source created with string_view don't copy (equiv to old SourceView) and Source created with std::string copies (equiv to old Source)

gmagogsfm replied:
In that case, `source` may or may not own the underlying string and that is dangerous. Could we preserve the original inheritance structure so that it is clear whether ownership is needed or not?

qihqi commented on torch/csrc/jit/frontend/source_range.h:
```diff
@@ -4,37 +4,320 @@
+};
+
+// the interface
+struct SourceView {
```
Good point, we don't need.

qihqi committed:
Make debug_pkl smaller by only emitting unique traces. (#72596)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/72596

debug_pkl file inside of pytorch's .pt file consists of a list of SourceRanges. Each SourceRange points to a Source which is a stack track, filename, and start, end numbers. Those are emitted in debug_pkl file as strings.

Since many SourceRange shares the same source, the string for trace can be deduped.

The newer format saves a set of unique traces in a tuple, then each SourceRange will save the offset of it's trace w.r.t. position in that tuple. (i.e. manually applying dictionary compression).

The above helps with smaller file size. On loading, if we copy each trace to Source as string the runtime memory would still blowup.
To mitigate this, we use SourceView directly instead of source which will take the reference of string inside of Deserializer and make that into string_view. This is safe because Deserializer is hold by Unpickler by shared_ptr, and Unpickler is also hold by shared_ptr by another Source object. That Source object will be alive during the model construction.

Test Plan:
unit test

Took original file (312271638_930.predictor.disagg.local); loaded with `torch.jit.load` save again with `torch.jit.save`. Unzip both, look at contents:
```
[qihan@devvm5585.vll0 ~]$ du archive -h
4.0K    archive/xl_model_weights
3.7M    archive/extra
8.0K    archive/code/__torch__/caffe2/torch/fb/model_transform/splitting
8.0K    archive/code/__torch__/caffe2/torch/fb/model_transform
8.0K    archive/code/__torch__/caffe2/torch/fb
8.0K    archive/code/__torch__/caffe2/torch
8.0K    archive/code/__torch__/caffe2
20M     archive/code/__torch__/torch/fx/graph_module
20M     archive/code/__torch__/torch/fx
8.0K    archive/code/__torch__/torch/classes
20M     archive/code/__torch__/torch
20M     archive/code/__torch__
20M     archive/code
2.7M    archive/constants
35M     archive
[qihan@devvm5585.vll0 ~]$ du resaved -h
4.0K    resaved/extra
8.0K    resaved/code/__torch__/caffe2/torch/fb/model_transform/splitting
8.0K    resaved/code/__torch__/caffe2/torch/fb/model_transform
8.0K    resaved/code/__torch__/caffe2/torch/fb
8.0K    resaved/code/__torch__/caffe2/torch
8.0K    resaved/code/__torch__/caffe2
1.3M    resaved/code/__torch__/torch/fx/graph_module
1.3M    resaved/code/__torch__/torch/fx
8.0K    resaved/code/__torch__/torch/classes
1.4M    resaved/code/__torch__/torch
1.4M    resaved/code/__torch__
1.4M    resaved/code
2.7M    resaved/constants
13M     resaved
[qihan@devvm5585.vll0 ~]$
```

Reviewed By: JasonHanwen

Differential Revision: D33994011

fbshipit-source-id: a6db7e1d4c616ced4442ec06936ae005bf269212
